### PR TITLE
tikv: implement pessimistic lock protocol

### PR DESCRIPTION
**Pull request by tiancaiamao**
PTAL @coocood

### CODE REVIEW

**Comment by coocood** [comment]
PLock should not block read, but seems there is no logic for it.

**Comment by coocood** [comment]
Prewrite should detect PLock written by self, and update it to prewrite lock.

File: tikv/mvcc.go [hunk 1]
```diff
@@ -144,6 +152,84 @@ func (store *MVCCStore) Prewrite(reqCtx *requestCtx, mutations []*kvrpcpb.Mutati
...
+	for _, m := range mutations {
+		lock, err := store.checkConflictInLockStore(reqCtx, m, startTS)
+		if err != nil {
+			anyError = true
+		}
+		if lock != nil {
+			if lock.Op == uint8(kvrpcpb.Op_PessimisticLock) {
+				// lockstore doesn't support update for now, delete the lock first.
+				lockBatch.delete(m.Key)
+				hasPessimistic = true
```
Comment (lysu) [inline tikv/mvcc.go]: maybe we should only set `hasPessimistic` flag just for in current `m`, other mutators also need check conflict?

Comment (coocood) [reply tikv/mvcc.go]: pessimistic lock never conflicts in prewrite.

File: tikv/mvcc.go [hunk 2]
```diff
@@ -130,11 +136,13 @@ func (store *MVCCStore) Prewrite(reqCtx *requestCtx, mutations []*kvrpcpb.Mutati
...
 	for _, m := range mutations {
-		duplicate, err := store.checkPrewriteInLockStore(reqCtx, m, startTS)
+		lock, err := store.checkConflictInLockStore(reqCtx, m, startTS)
 		if err != nil {
 			anyError = true
 		}
-		if duplicate {
+		if lock != nil {
+			// The lock already exists, it's a duplicate command, we can return directly.
+			y.Assert(lock.Op == uint8(kvrpcpb.Op_PessimisticLock))
```
Comment (tiancaiamao) [inline tikv/mvcc.go]: This could be done in TiDB too, and the performance should be better.

File: tikv/mvcc.go [hunk 3]
```diff
@@ -229,14 +324,14 @@ func (store *MVCCStore) checkPrewriteInLockStore(
...
-func (store *MVCCStore) checkPrewriteInDB(
+func (store *MVCCStore) checkConflictInDB(
 	req *requestCtx, txn *badger.Txn, item *badger.Item, startTS uint64) (userMeta mvcc.DBUserMeta, val []byte, err error) {
 	if item == nil {
 		return nil, nil, nil
 	}
 	userMeta = mvcc.DBUserMeta(item.UserMeta())
 	if userMeta.CommitTS() > startTS {
-		return nil, nil, ErrRetryable("write conflict")
+		return nil, nil, ErrRetryable("[write conflict]")
```
Comment (tiancaiamao) [inline tikv/mvcc.go]: This kind of code is fragile IMHO

File: tikv/mvcc.go [hunk 4]
```diff
@@ -144,6 +152,84 @@ func (store *MVCCStore) Prewrite(reqCtx *requestCtx, mutations []*kvrpcpb.Mutati
...
+			anyError = true
+		}
+		if lock != nil {
+			if lock.Op == uint8(kvrpcpb.Op_PessimisticLock) {
+				// lockstore doesn't support update for now, delete the lock first.
+				lockBatch.delete(m.Key)
+				hasPessimistic = true
+			} else {
+				// duplicated command.
+				return nil
```
Comment (tiancaiamao) [inline tikv/mvcc.go]: Should it be continue here?

File: tikv/mvcc.go (hunk 2 above)
Comment (tiancaiamao) [reply tikv/mvcc.go]: Done

File: tikv/mvcc.go [hunk 5]
```diff
@@ -119,22 +129,108 @@ func (store *MVCCStore) updateLatestTS(ts uint64) {
...
 	// Must check the LockStore first.
-	for _, m := range mutations {
-		duplicate, err := store.checkPrewriteInLockStore(reqCtx, m, startTS)
+	for i, m := range mutations {
+		lock, err := store.checkConflictInLockStore(reqCtx, m, req.StartVersion)
 		if err != nil {
 			anyError = true
 		}
-		if duplicate {
+		if hasPessimistic && req.PessimisticLockExist[i] {
```
Comment (youjiali1995) [inline tikv/mvcc.go]: What's the use of req.PessimisticLockExist[i]? Only all pessimistic locks exist, prewrite can succeed.

Comment (coocood) [reply tikv/mvcc.go]: non-unique index and delete index don't need to be locked.

File: util/lockwaiter/lockwaiter.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	Position Position
+	CommitTS uint64
+}
+
+const WaitTimeout Position = -1
+
+func (w *Waiter) Wait() Result {
+	select {
+	case <-time.After(w.timeout):
+		return Result{Position: WaitTimeout}
```
Comment (youjiali1995) [inline util/lockwaiter/lockwaiter.go]: When timeout, the waiter should clean up itself.

**Comment by coocood** [comment]
replaced by https://github.com/ngaut/unistore/pull/202

**Pull request closed without merging**